crates/farming/src/lib.rs

```
@@ -0,0 +1,357 @@
+                &pool_account_id,
+                &treasury_account_id,
+                T::MultiCurrency::free_balance(reward_currency_id, &pool_account_id),
+            )?;
+
+            RewardSchedules::<T>::remove(pool_currency_id, reward_currency_id);
+            Self::deposit_event(Event::RewardScheduleUpdated {
+                pool_currency_id,
+                reward_currency_id,
+                period_count: Zero::zero(),
+                per_period: Zero::zero(),
+            });
+
+            Ok(().into())
+        }
+
+        /// Stake the pool tokens in the reward pool
+        #[pallet::weight(T::WeightInfo::deposit())]
+        #[transactional]
+        pub fn deposit(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // prevent depositing without reward schedule
+            ensure!(
+                !RewardSchedules::<T>::iter_prefix_values(pool_currency_id)
+                    .count()
+                    .is_zero(),
+                Error::<T>::ScheduleNotFound
+            );
+
+            // reserve lp tokens to prevent spending
+            T::MultiCurrency::reserve(pool_currency_id.clone(), &who, amount)?;
+
+            // deposit lp tokens as stake
+            T::RewardPools::deposit_stake(&pool_currency_id, &who, amount)
+        }
+
+        /// Unstake the pool tokens from the reward pool
+        #[pallet::weight(T::WeightInfo::withdraw())]
+        #[transactional]
+        pub fn withdraw(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // unreserve lp tokens to allow spending
+            let _remaining = T::MultiCurrency::unreserve(pool_currency_id.clone(), &who, amount);
```

Should we not catch this error? I mean, it if withdraw_stake succeeds, then unreserving _should_ succeeds as well, but I see no good reason to explicitly ignore any errors

---

The only time `_remaining` would be non-zero is if `amount > reserved` but since `withdraw_stake` will anyway limit the total amount that can be withdrawn I think we can ignore this check

---

Better safe than sorry?

---

Huh I assumed `unreserve` would return a `Result`, but actually that's not the case. With your updated version you're still ignoring the returned value, technically we should check that the returned value is zero